File: leetcode/125.py
"""
    File: 125.py
    Title: Valid Palindrome
    Difficulty: Easy
"""


class Solution:
    def isPalindrome(self, s: str) -> bool:
        lo, hi = 0, len(s) - 1

        def valid(c):
            return "a" <= c <= "z" or "A" <= c <= "Z" or "0" <= c <= "9"

        while lo < hi:
            while lo < len(s) and not valid(s[lo]):
                lo += 1
            while hi > 0 and not valid(s[hi]):
                hi -= 1
            if lo >= len(s) or hi < 0:
                break
            if s[lo].lower() != s[hi].lower():
                return False
            lo += 1
            hi -= 1
        return True

File: leetcode/125.go
// File: 125.go
// Title: Valid Palindrome
// Difficulty: Easy

package leetcode

import (
	"strings"
	"unicode"
)

func isPalindrome(s string) bool {
	alphanumeric := ""
	for _, r := range s {
		if unicode.IsLetter(r) || unicode.IsNumber(r) {
			alphanumeric += strings.ToLower(string(r))
		}
	}
	length := len(alphanumeric)
	for i := 0; i < length/2; i++ {
		if alphanumeric[i] != alphanumeric[length-i-1] {
			return false
		}
	}
	return true
}
